web3_proxy/src/frontend/users/referral.rs

```diff
@@ -70,3 +75,132 @@ pub async fn user_referral_link_get(
+            referral_start_date: referral_record.referral_start_date,
+            used_referral_code: referrer_record.referral_code,
+        };
+        // Start inserting json's into this
+        out.push(tmp);
+    }
+
+    // Turn this into a response
+    let response_json = json!({
+        "referrals": out,
+        "user": user,
+    });
+
+    let response = (StatusCode::OK, Json(response_json)).into_response();
+    Ok(response)
+}
+
+#[debug_handler]
+pub async fn user_shared_referral_stats(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    TypedHeader(Authorization(bearer)): TypedHeader<Authorization<Bearer>>,
+    Query(_params): Query<HashMap<String, String>>,
+) -> Web3ProxyResponse {
+    // First get the bearer token and check if the user is logged in
+    let (user, _semaphore) = app.bearer_is_authorized(bearer).await?;
+
+    let db_replica = app
+        .db_replica()
+        .context("getting replica db for user's revert logs")?;
+
+    // Get all referral records associated with this user
+    let referrals = referrer::Entity::find()
+        .filter(referrer::Column::UserId.eq(user.id))
+        .find_also_related(referee::Entity)
+        .all(db_replica.conn())
+        .await?;
+
+    // Return early if the user does not have any referred entities
+    if referrals.len() == 0 {
+        let response_json = json!({
+        "referrals": Vec::<()>::new(),
+        "used_referral_code": "",
+        "user": user,
+        });
+
+        let response = (StatusCode::OK, Json(response_json)).into_response();
+        return Ok(response);
+    }
+
+    // For each related referral person, find the corresponding user-address
+    #[derive(Debug, Serialize)]
+    struct Info {
+        credits_applied_for_referee: bool,
+        credits_applied_for_referrer: Decimal, // Probably make this a float instead (approximate anyways)
+        referral_start_date: DateTime,
+        referred_address: String, // Address of the referrer
+    };
+
+    let mut out: Vec<Info> = Vec::new();
+    let mut used_referral_code = "".to_owned(); // This is only for safety purposes, because of the condition above we always know that there is at least one record
+    for x in referrals.into_iter() {
+        let (referrer, referral_record) = (x.0, x.1.unwrap());
+        used_referral_code = referrer.referral_code;
+        // The foreign key is never optional
+        let referred_user = user::Entity::find_by_id(referral_record.user_id)
+            .one(db_replica.conn())
+            .await?
+            .context("Database error, no foreign key found for referring user")?;
+        let tmp = Info {
+            credits_applied_for_referee: referral_record.credits_applied_for_referee,
+            credits_applied_for_referrer: referral_record.credits_applied_for_referrer,
+            referral_start_date: referral_record.referral_start_date,
+            referred_address: format!("{:?}", Address::from_slice(referred_user.address.as_ref())),
```

Comment:
no need for the `format!`. `Address` already serializes how you want it

Reply:
I also mentioned this on another PR. Is it necessary to expose the on-chain address here? I'd prefer to keep that private as much as possible.

Reply:
I removed it. Was assuming that the referrer would know who to share the address with. Actually, he does know when making the referral, so I'll keep this in. Let me know if I should remove it though.